File: src/lib/clean-name.ts
/**
 * Titel-normalisatie voor scrape-resultaten.
 *
 * Vooral Bol-titels zijn vaak machinevertaald of door marketplace-verkopers
 * verminkt: "Wees Stil! ... Cpu-Koeler Voor Am4-Moederbord" (be quiet!),
 * "ATX Semi-tower Box Fractal North", "Amd ... 3,50/5,00 Ghz", of een twee
 * keer herhaalde productnaam. Deze module maakt daar leesbare titels van.
 *
 * Spiegel van scrapers/corebuild_scrapers/clean_name.py — bij wijzigingen
 * beide bestanden bijwerken.
 */

/** Verminkte merknamen (letterlijke vertalingen e.d.). */
const BRAND_FIXES: [RegExp, string][] = [
  [/\bwees stil!?/gi, "be quiet!"],
];

/** Beschrijvende categorie-prefixen die vóór de echte productnaam staan. */
const PREFIX_STRIP =
  /^(atx (semi-?tower )?box|box ventilator( pc)?|motherboard|processor|cpu -)\s+/i;

/** Vertaalde/nietszeggende marketing-fragmenten. */
const NOISE_FRAGMENTS: RegExp[] = [
  /\bbehuizing voor s-?am\d+ computer\b/gi, // "Box for S-AM5 computer" = boxed
  /,?\s*hoge prestaties\.?$/gi,
];

/** Correcte schrijfwijze van veelvoorkomende termen (case-insensitive match). */
const CASING: Record<string, string> = {
  amd: "AMD",
  intel: "Intel",
  ryzen: "Ryzen",
  nvidia: "NVIDIA",
  geforce: "GeForce",
  radeon: "Radeon",
  asus: "ASUS",
  msi: "MSI",
  ghz: "GHz",
  mhz: "MHz",
  cpu: "CPU",
  gpu: "GPU",
  psu: "PSU",
  ddr3: "DDR3",
  ddr4: "DDR4",
  ddr5: "DDR5",
  nvme: "NVMe",
  rtx: "RTX",
  gtx: "GTX",
  atx: "ATX",
  wifi: "WiFi",
  usb: "USB",
  hdmi: "HDMI",
  ssd: "SSD",
  hdd: "HDD",
  argb: "ARGB",
  rgb: "RGB",
  pwm: "PWM",
  aio: "AIO",
  oem: "OEM",
  am4: "AM4",
  am5: "AM5",
  lga: "LGA",
  pcie: "PCIe",
};

/** Verwijder een direct herhaalde woordreeks ("Fractal North Fractal North"). */
function dedupePhrases(name: string): string {
  const tokens = name.split(/\s+/);
  for (let len = Math.floor(tokens.length / 2); len >= 1; len--) {
    for (let i = 0; i + 2 * len <= tokens.length; i++) {
      const a = tokens.slice(i, i + len).join(" ").toLowerCase();
      const b = tokens.slice(i + len, i + 2 * len).join(" ").toLowerCase();
      if (a === b) {
        tokens.splice(i + len, len);
        return dedupePhrases(tokens.join(" "));
      }
    }
  }
  return tokens.join(" ");
}

export function cleanName(raw: string): string {
  let name = raw.replace(/\s+/g, " ").trim();

  for (const [re, fix] of BRAND_FIXES) name = name.replace(re, fix);
  name = name.replace(PREFIX_STRIP, "");
  for (const re of NOISE_FRAGMENTS) name = name.replace(re, "");

  // Schrijfwijze per term, alleen op woordgrenzen ("Ghz" → "GHz", "Cpu" → "CPU")
  for (const [key, value] of Object.entries(CASING)) {
    name = name.replace(new RegExp(`\\b${key}\\b`, "gi"), value);
  }
  // "120Mm" → "120mm", "Rx 7600" → "RX 7600" (alleen met cijfers erachter)
  name = name.replace(/\b(\d+)\s?mm\b/gi, "$1mm");
  name = name.replace(/\brx(?=\s?\d{3,4})/gi, "RX");

  name = dedupePhrases(name);

  // Losse scheidingstekens en restkomma's opruimen
  name = name
    .replace(/\s+([,.])/g, "$1")
    .replace(/,\s*,/g, ",")
    .replace(/\s{2,}/g, " ")
    .replace(/[\s,.\-–|]+$/g, "")
    .trim();

  return name.length >= 3 ? name : raw.trim();
}

File: scrapers/corebuild_scrapers/clean_name.py
"""Titel-normalisatie voor scrape-resultaten.

Vooral Bol-titels zijn vaak machinevertaald of door marketplace-verkopers
verminkt. Spiegel van src/lib/clean-name.ts - bij wijzigingen beide
bestanden bijwerken.
"""

import re

_BRAND_FIXES: list[tuple[re.Pattern, str]] = [
    (re.compile(r"\bwees stil!?", re.I), "be quiet!"),
]

_PREFIX_STRIP = re.compile(
    r"^(atx (semi-?tower )?box|box ventilator( pc)?|motherboard|processor|cpu -)\s+", re.I
)

_NOISE_FRAGMENTS: list[re.Pattern] = [
    re.compile(r"\bbehuizing voor s-?am\d+ computer\b", re.I),
    re.compile(r",?\s*hoge prestaties\.?$", re.I),
]

_CASING: dict[str, str] = {
    "amd": "AMD", "intel": "Intel", "ryzen": "Ryzen", "nvidia": "NVIDIA",
    "geforce": "GeForce", "radeon": "Radeon", "asus": "ASUS", "msi": "MSI",
    "ghz": "GHz", "mhz": "MHz", "cpu": "CPU", "gpu": "GPU", "psu": "PSU",
    "ddr3": "DDR3", "ddr4": "DDR4", "ddr5": "DDR5", "nvme": "NVMe",
    "rtx": "RTX", "gtx": "GTX", "atx": "ATX", "wifi": "WiFi", "usb": "USB",
    "hdmi": "HDMI", "ssd": "SSD", "hdd": "HDD", "argb": "ARGB", "rgb": "RGB",
    "pwm": "PWM", "aio": "AIO", "oem": "OEM", "am4": "AM4", "am5": "AM5",
    "lga": "LGA", "pcie": "PCIe",
}


def _dedupe_phrases(name: str) -> str:
    tokens = name.split()
    for length in range(len(tokens) // 2, 0, -1):
        for i in range(len(tokens) - 2 * length + 1):
            a = " ".join(tokens[i : i + length]).lower()
            b = " ".join(tokens[i + length : i + 2 * length]).lower()
            if a == b:
                del tokens[i + length : i + 2 * length]
                return _dedupe_phrases(" ".join(tokens))
    return " ".join(tokens)


def clean_name(raw: str) -> str:
    name = re.sub(r"\s+", " ", raw).strip()

    for pattern, fix in _BRAND_FIXES:
        name = pattern.sub(fix, name)
    name = _PREFIX_STRIP.sub("", name)
    for pattern in _NOISE_FRAGMENTS:
        name = pattern.sub("", name)

    for key, value in _CASING.items():
        name = re.sub(rf"\b{key}\b", value, name, flags=re.I)
    name = re.sub(r"\b(\d+)\s?mm\b", r"\1mm", name, flags=re.I)
    name = re.sub(r"\brx(?=\s?\d{3,4})", "RX", name, flags=re.I)

    name = _dedupe_phrases(name)

    name = re.sub(r"\s+([,.])", r"\1", name)
    name = re.sub(r",\s*,", ",", name)
    name = re.sub(r"\s{2,}", " ", name)
    name = re.sub(r"[\s,.\-–|]+$", "", name).strip()

    return name if len(name) >= 3 else raw.strip()
